Implement singularity-compose.yml override configuration feature

PauloMigAlmeida opened the pull request:
There is one feature in docker-compose that I would love to have in singularity-compose that is the ability to specify multiple composer.yml files in which the properties would get either merged or overidden.

Example:

singularity-compose.yml
```YAML
version: "2.0"
instances:

  cvat:
    image: cvat.sif
    start:
      options:
        - containall
    volumes:
      - ./recipes/cvat/env.sh:/.singularity.d/env/env.sh
      - ./volumes/cvat/data:/home/django/data
      - ./volumes/cvat/keys:/home/django/keys
      - ./volumes/cvat/logs:/home/django/logs
      - ./volumes/cvat/logs:/var/log/supervisor
      - ./volumes/cvat/tmp:/tmp
    network:
      enable: false

```

singularity-compose-dev.yml
```YAML
version: "2.0"
instances:

  cvat:
    start:
      options:
        - containall
        - fakeroot

  cvatproxy:
    image: cvatproxy.sif
    start:
      options:
        - containall
    volumes:
      - ./volumes/cvatproxy/cache/nginx:/var/cache/nginx
      - ./volumes/cvatproxy/run:/var/run/
      - ./volumes/cvatproxy/log:/var/log/nginx
      - ./volumes/cvatproxy/lib:/var/lib/nginx
    network:
      enable: false
    depends_on:
      - cvat
```

By executing something like:

```
singularity-compose -f singularity-compose.yml -f singularity-compose.dev.yml <any_command>
```

we would get the runtime config combined into:

```YAML
version: "2.0"
instances:

  cvat:
    image: cvat.sif
    start:
      options:
        - containall
        - fakeroot
    volumes:
      - ./recipes/cvat/env.sh:/.singularity.d/env/env.sh
      - ./volumes/cvat/data:/home/django/data
      - ./volumes/cvat/keys:/home/django/keys
      - ./volumes/cvat/logs:/home/django/logs
      - ./volumes/cvat/logs:/var/log/supervisor
      - ./volumes/cvat/tmp:/tmp
    network:
      enable: false

  cvatproxy:
    image: cvatproxy.sif
    start:
      options:
        - containall
    volumes:
      - ./volumes/cvatproxy/cache/nginx:/var/cache/nginx
      - ./volumes/cvatproxy/run:/var/run/
      - ./volumes/cvatproxy/log:/var/log/nginx
      - ./volumes/cvatproxy/lib:/var/lib/nginx
    network:
      enable: false
    depends_on:
      - cvat
```

which essentially overrode cvat->start->options and appended/merged cvatproxy to the runtime config.

I'm looking for feedback on the following points

1. whether or not you agree this is a feature that you want to see in singularity-compose?
2. I did a draft implementation, any thing that you want me to change or implement differently?

vsoch commented on scompose/project/project.py:
```diff
@@ -64,8 +64,8 @@ def set_filename(self, filename):
         """
-        self.filename = filename or "singularity-compose.yml"
-        self.working_dir = os.path.dirname(os.path.abspath(self.filename))
+        self.filename = filename
```
I would leave the default here to "singularity-compose.yml" - and if you are changing this to require a list, then please support the old behavior to still provide a single filename, and have it checked if it's a list, single string, or None, and process accordingly. And then if your preference is to store in a list, the variable should be self.filenames to indicate plural / multiple files. 

And if I understand correctly, the working directory is changed to os.getcwd() to account for possibly different compose files in different directories. Of course the issue here is that all compose files need to then have their context relevant to that directory (is that your intention?)

vsoch replied:
yes correct! We wouldn't want to break previous scripts that are providing a string. The simple solution I think is to allow either.

vsoch commented on scompose/project/project.py:
```diff
@@ -148,15 +145,59 @@ def get_already_running(self):
+            for f in self.filename:
+                # ensure file exists
+                if not os.path.exists(f):
+                    bot.error("%s does not exist." % f)
```
Please move the multiple load of yaml files outside of the project.py - there should ideally be a separate utils function or set of config functions (some config.py or similar) that takes the list. and returns the single config here. Also - self.filename implies one file, and there should either be a self.filenames or self.filename to match the arity.

vsoch commented on scompose/project/project.py:
```diff
@@ -148,15 +145,59 @@ def get_already_running(self):
         except:  # ParserError
             bot.exit("Cannot parse %s, invalid yaml." % self.filename)
 
+    def deep_merge(self, yaml_files):
+        """merge singularity-compose.yml files into a single dict"""
+        if len(yaml_files) == 1:
+            # nothing to merge as the user specified a single file
+            return yaml_files[0]
+
+        base_yaml = None
+        for idx, item in enumerate(yaml_files):
+            if idx == 0:
+                base_yaml = item
+            else:
+                base_yaml = self.merge(base_yaml, item)
+
+        return base_yaml
+
```
All of this logic should not be in project.py - it's for merging yaml.

vsoch commented on scompose/project/project.py:
```diff
@@ -148,15 +145,59 @@ def get_already_running(self):
+
+        return base_yaml
+
+    def merge(self, a, b):
```
Move this too.

vsoch reviewed (commented):
Very cool idea! We could definitely support this. I've added some comments because we need to:

 - support previous functionality (e.g., the user providing no -f defaults to a single singularity-compose.yml as before)
 - the Python client in project.py also needs to work the same, still accepting a string (and you are extending to allow a list to merge).
 - The merge / load multiple yaml should not live in project.py - you can create a new module for it until utils if that's appropriate!

vsoch commented:
Also - if we provide this functionality I'd add an ability to preview a merge, basically taking the files, doing the merge, and printing yaml back to the screen.

PauloMigAlmeida commented:
> Also - if we provide this functionality I'd add an ability to preview a merge, basically taking the files, doing the merge, and printing yaml back to the screen.

Do you mean like a CLI option similar to what `--read_only` is ? Or printing it to console with (let's debugging logging verbosity)?

vsoch commented:
> Do you mean like a CLI option similar to what --read_only is ? Or printing it to console with (let's say logging verbosity set to debug)?

It could be that we have a separate command group for the config, and then something like a `--preview`?  It doesn't have to match docker-compose exactly, although if docker compose already has that functionality we could. E.g.,:

```
# If we still use the up command group
$ singularity-compose up -d --preview

# A more general validation function (that could validate...
$ singularity-compose config --check

# And then support a preview
$ singularity-compose config --check
```
We can definitely split this into two different PRs - the reason I mentioned it is because not having the final file in a repository, for example, hinders reproducibility. So minimally I think we'd want to be able to output it to slightly better encourage saving it :)

PauloMigAlmeida commented:
Hi @vsoch, apologies for asking one more question on that subject, I just want to make sure I understood what needs to be done.

Currently, if we run `singularity-compose -f file1 -f file2 config` this will already print the interpolated version (if multiple files are specified) or the original file (if just a single one is specified).  The only caveat with that command is that it prints the content in JSON instead of YAML but given that its contents is interchangeable then that all good.

(that's the part which I'm not 100% sure if I got right, please correct me where I'm wrong):
If I understood you correctly, you want `singularity-compose config --check` to do pretty much the same thing but printing it in YAML format so that would slightly encourage devs to save the content. 

If the above is not correct, I know sure if I got the `--check` option's purpose 😬

PauloMigAlmeida commented:
thanks for the clarification, I got it know :) 

Sure, feel free to commit stuff to this PR. For now, I will focus on adding the tests and docs related to this feature before taking on any of the other stuff.

vsoch commented:
Sounds good! And I can definitely help too - I've done a bit of work with jsonschema so it's pretty quick for me to do.

PauloMigAlmeida commented:
Your help is always welcome 😃

PauloMigAlmeida commented:
I moved the logic out of project, added tests and docs. I will mark this PR as 'ready to review' just in case you want to expand those functionalities in a different PR.

vsoch commented:
Here is what I had in mind for check! https://github.com/singularityhub/singularity-compose/pull/49

So if we add a `--preview` to that given two combined files (as you've done here) we can print the final thing to the console.

vsoch commented:
And I am aware of the comedic value of failing my own lint check! I think it's probably time to remove the version constraint for black, and assume that most people will create a new venv to have it installed freshly. :laughing:

PauloMigAlmeida commented:
A bit off-topic (in relation to this PR) but something that it would be nice for us to start thinking before writing any code... another thing I would love to have on singularity-compose is to be able to create volumes and reference them by name rather than path.

Usually docker create these persistent volumes under `$HOME/.docker`  and I was thinking about using a path along the same train of thought. 

I cannot count how many times I've (accidentally) deleted my 'persistent' paths before as I'm always tempted to have those mapped folders within my project directory structure 😬  There are other cases in which this is relevant (such as file system capabilities) but the key point is, if we could start thinking about this now, we can get a nice implementation shortly 😄

PauloMigAlmeida commented:
Moved that to an issue to make sure the 2 subjects won't get mixed (I should've thought about that before 😓 )

vsoch commented:
This is such a great feature - thank you @PauloMigAlmeida !
